1DRO Unit Model

adam-a-a opened the pull request:
## Fixes/Addresses: #126 

## Summary/Motivation:
Complete the 1DRO unit model with all detailed calculations included. Any feedback on improving functionality where possible would be much appreciated. 

Additionally, some of the newest variables in the 0DRO model (e.g., over pressure ratio) were not added in this PR; those variables can be added to this PR or a subsequent one based on reviewer feedback.

## Changes proposed in this PR:
- add all config options for concentration polarization, mass transfer coefficient, and pressure drop
- add associated testing
-  adjust initialization method and associated testing in 0dro along the way
-  add `has_full_reporting` config option and methods to compute expressions for averaged variables that will show up in a detailed report of results (via `report()`)

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

TimBartholomew reviewed (commented):
@adam-a-a does this include a verification test to my published work?

adam-a-a commented:
> @adam-a-a does this include a verification test to my published work?

@TimBartholomew I modified the first big test (TestReverseOsmosis) to match the fixed values from your previous study; the results are close although the membrane area that I found was ~ 22 m^2 while your study showed 19 m^2. My assumption is that this deviation could be due to the adjustment made to the osmotic pressure calculation in our NaCl model.

TimBartholomew commented:
> @TimBartholomew I modified the first big test (TestReverseOsmosis) to match the fixed values from your previous study; the results are close although the membrane area that I found was ~ 22 m^2 while your study showed 19 m^2. My assumption is that this deviation could be due to the adjustment made to the osmotic pressure calculation in our NaCl model.

So the flux difference is 10-15%? That seems a little high to be just the osmotic pressure calculation because I think the previous error was pronounced at high salinities and that the error was less than 5% at seawater/RO concentrate salinities. We can check what the difference is at those salinities, as well as run that specific test with the old osmotic pressure calculation. 

How do the other things compare in Table 2 (i.e pressure drop, salt passage, average Reynolds number, average mass transfer coefficient)?

adam-a-a commented:
> > @TimBartholomew I modified the first big test (TestReverseOsmosis) to match the fixed values from your previous study; the results are close although the membrane area that I found was ~ 22 m^2 while your study showed 19 m^2. My assumption is that this deviation could be due to the adjustment made to the osmotic pressure calculation in our NaCl model.
> 
> So the flux difference is 10-15%? That seems a little high to be just the osmotic pressure calculation because I think the previous error was pronounced at high salinities and that the error was less than 5% at seawater/RO concentrate salinities. We can check what the difference is at those salinities, as well as run that specific test with the old osmotic pressure calculation.
> 
> How do the other things compare in Table 2 (i.e pressure drop, salt passage, average Reynolds number, average mass transfer coefficient)?

So, I adjusted some of the inputs (fixed vars) to more closely match yours. Furthermore, I computed the average Re number, average mass transfer coefficient, and average fluxes. The flux values that I had before were values at the outlet. Below are some values from your study that I tested against. The results are quite close. For average water flux, % difference is ~6.25% compared to your study's result.

- Pressure drop, study = 1.5 bar | Pressure drop,model = 1.56 bar
- Average water flux, study = 25.6 LMH | Average water flux,model = 24 LMH
- Average salt flux, study = 8.1 grams/m2/h | Average salt flux,model = 8.0 grams/m2/h
- Average Reynolds number, study = 272 | Average Reynolds number,model = 274
- Pressure drop, study = 1.5 bar | Pressure drop,model = 1.56 bar

TimBartholomew commented on proteuslib/unit_models/tests/test_reverse_osmosis_1D.py:
```diff
@@ -293,4 +565,843 @@ def test_solution(self, RO_frame):
+        m.fs.unit.recovery_vol_phase[0, 'Liq'].fix(0.4)
+        m.fs.unit.cp_modulus.fix(1.1)
+
+        # test ports
```
We don't need to do this for each case, I'd get rid of this for all but the first one.

TimBartholomew commented on proteuslib/unit_models/tests/test_reverse_osmosis_1D.py:
```diff
@@ -293,4 +565,843 @@ def test_solution(self, RO_frame):
+            assert len(port.vars) == 3  # number of state variables for NaCl property package
+            assert isinstance(port, Port)
+
+        # test pyomo objects on unit
```
We don't need to do this for each case, I'd get rid of this for all but the first one.

adam-a-a replied:
I decided to do this since different cases can have a slightly different set of variable/constraint objects. So while this may seem exhaustive, I suppose it does serve a purpose.

TimBartholomew replied:
Yeah but then I would just check the things that are different from the main detailed case. Regardless all of this will be reduced with a unit test harness that hopefully I do next week.

TimBartholomew commented on proteuslib/unit_models/tests/test_reverse_osmosis_1D.py:
```diff
@@ -293,4 +565,843 @@ def test_solution(self, RO_frame):
+                continue  # do not test hidden references
+            assert obj_str in unit_objs_type_dict
+
+        # test feed-side control volume and associated stateblocks
```
We don't need to do this for each case, I'd get rid of this for all but the first one.

TimBartholomew commented on proteuslib/unit_models/tests/test_reverse_osmosis_1D.py:
```diff
@@ -293,4 +565,843 @@ def test_solution(self, RO_frame):
+            sb = getattr(m.fs.unit.feed_side, sb_str)
+            assert isinstance(sb, props.NaClStateBlock)
+
+        stateblock_lst = ['permeate_side', 'permeate_out']
```
We don't need to do this for each case, I'd get rid of this for all but the first one.

TimBartholomew commented on proteuslib/unit_models/tests/test_reverse_osmosis_1D.py:
```diff
@@ -293,4 +565,843 @@ def test_solution(self, RO_frame):
+        assert (abs(value(flow_mass_inlet - flow_mass_retentate - flow_mass_permeate))
+                <= 1e-2)
+
+        # Test solution
```
We should also test the total pressure drop for the cases. And the inlet flux.

adam-a-a replied:
Will do. Just a note- the "inlet" flux (first flux computed) is located one index after `length_domain.first()` since the constraint is skipped at the actual inlet. Does that make sense?

TimBartholomew reviewed (commented):
Just minor comments on testing. I think all of the testing for the individuals cases can be deleted.

TimBartholomew commented:
> So, I adjusted some of the inputs (fixed vars) to more closely match yours. Furthermore, I computed the average Re number, average mass transfer coefficient, and average fluxes. The flux values that I had before were values at the outlet. Below are some values from your study that I tested against. The results are quite close. For average water flux, % difference is ~6.25% compared to your study's result.
> 
> * Pressure drop, study = 1.5 bar | Pressure drop,model = 1.56 bar
> * Average water flux, study = 25.6 LMH | Average water flux,model = 24 LMH
> * Average salt flux, study = 8.1 grams/m2/h | Average salt flux,model = 8.0 grams/m2/h
> * Average Reynolds number, study = 272 | Average Reynolds number,model = 274
> * Average mass transfer coefficient, study = 113 mm/h | Average mass transfer coefficient , model = 112 mm/h

Ok that looks pretty good to me.

TimBartholomew reviewed (dismissed):
LGTM. Noting that we are going to be refactoring both the RO models and the unit model test in the near future :)

bknueven commented on proteuslib/unit_models/reverse_osmosis_1D.py:
```diff
@@ -563,49 +1029,118 @@ def initialize(blk,
         opt = get_solver(solver, optarg)
 
-        init_log.info('Starting initialization')
+        init_log.info('Starting Initialization Step 1: initialize blocks.')
```
The 0D-RO unit model puts most of its initialization output in `info_high`, except to report success.

adam-a-a replied:
I changed this because the messages were not being displayed when I set `info_high`.

bknueven replied:
This snippet `ro_unit.initialize(outlvl=ideaslog.INFO_HIGH)` gives be the `info_high` initialization logs on the 0D-RO model.

Either way is perfectly valid; just pointing out an inconsistency.

bknueven reviewed (approved):
This seems reasonable. I left a nit, but it can be saved, especially given both RO models are due for refactoring.

adam-a-a commented:
@TimBartholomew @bknueven  just noticed that several tests are being skipped locally so want to fix that before merging. I'll re-request your reviews soon =)

adam-a-a commented:
@TimBartholomew @bknueven @andrewlee94 This PR should be ready to go.

TimBartholomew reviewed (approved):
LGTM
